src/cryptonote_core/service_node_list.cpp

```diff
@@ -312,7 +320,114 @@ namespace service_nodes
+    if (count == MIN_SWARM_SIZE - 1)
+    {
+      // XXX XXX XXX XXX
+      // All nodes should fetch data from this swarm at this point only.
+      // The registered nodes will eventually disappear and after this point,
+      // it is already considered gone. It only exists to retrieve data from.
+      //
+      // XXX XXX XXX XXX
+      //         internship optimization hardfork idea:
+      //         when nodes have been decomissioned for more than 10 blocks,
+      //         move the nodes into new swarms immediately.
+    }
+  }
+
+  uint64_t service_node_list::get_new_node_swarm_id(uint64_t block_height, uint32_t index) const
+  {
+    // XXX XXX XXX improvement proposal: have the swarm id be "remembered" from
+    // the last time, if the new node was already registered before. This would
+    // have huge performance gains preventing nodes from having to resync every
+    // time they rejoin the network.
+
+    crypto::hash hash = m_blockchain.get_block_id_by_height(block_height);
+    uint64_t seed = 0;
+    std::memcpy(&seed, hash.data, sizeof(seed));
+    seed += index;
+    std::mt19937_64 mersenne_twister(seed);
+
+    std::vector<uint64_t> swarm_ids = get_swarm_ids();
+
+    if (swarm_ids.empty())
+      return 0;
+
+    const size_t swarm_index = (size_t)uniform_distribution_portable(mersenne_twister, swarm_ids.size());
+
+    return swarm_ids[swarm_index];
+  }
+
+  void service_node_list::check_and_process_swarm_overflow(uint64_t swarm_id, uint64_t block_height)
+  {
+    uint64_t count = 0;
+    for (const auto& entry : m_service_nodes_infos)
+      if (entry.second.swarm_id == swarm_id)
+        count++;
+
+    if (count > MAX_SWARM_SIZE)
+    {
+      crypto::hash hash = m_blockchain.get_block_id_by_height(block_height);
+      uint64_t seed = 0;
+      std::memcpy(&seed, hash.data, sizeof(seed));
+      seed += swarm_id;
+      std::mt19937_64 mersenne_twister(seed);
+
+      bool move_to_new_swarm = (bool)uniform_distribution_portable(mersenne_twister, 2);
+      uint64_t new_swarm = uniform_distribution_portable(mersenne_twister, UINT64_MAX);
+      for (auto& entry : m_service_nodes_infos)
+      {
+        if (entry.second.swarm_id == swarm_id)
+        {
+          if (move_to_new_swarm)
+          {
+            m_rollback_events.push_back(std::unique_ptr<rollback_event>(new rollback_change(block_height, entry.first, entry.second)));
+            entry.second.swarm_id = new_swarm;
+          }
+          move_to_new_swarm = !move_to_new_swarm;
+        }
+      }
+    }
+
+    // XXX NOTE FOR DATA MIGRATION DEVS:
+    // nodes that are members of new_swarm should now query every
+    // single service node for data that belongs to new_swarm.
+
+    // XXX intern task for speed improvements.
+    // Use http://mathworld.wolfram.com/VoronoiDiagram.html to calculate exactly
+    // which service nodes to query (doesn't have to query all)
+  }
+
+  std::vector<uint64_t> service_node_list::get_swarm_ids() const
```

Comment:
Returning a vector of swarm_id and count looks to be useful here.

Reply:
you aren't going to need it.

if you (or somebody) needs it then put it in later.

Reply:
The rationale is, you use the count of a swarm multiple times, so, seems logical.

Reply:
ah yeah fair enough

Reply:
but I don't need the count of every swarm, so I've made a function to get the swarm size for a single swarm